File: packages/template-ts/tests/conformance.test.ts
// Conformance cases rendered in-process (CNF-6 to CNF-9).
import { existsSync, readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { describe, expect, it } from 'vitest';
import { AstProgram, BindError, Engine, parse, parseJsonBytes, TemplateError } from '../src/index.js';
import { FsLoader } from '../src/node/index.js';
import { casesDir, toJsonValue } from './helpers.js';

interface Case {
  id: string;
  dir: string;
}

function listCases(): Case[] {
  const cases: Case[] = [];
  for (const group of readdirSync(casesDir, { withFileTypes: true })) {
    if (!group.isDirectory()) continue;
    for (const entry of readdirSync(join(casesDir, group.name), { withFileTypes: true })) {
      if (entry.isDirectory()) cases.push({ id: `${group.name}/${entry.name}`, dir: join(casesDir, group.name, entry.name) });
    }
  }
  return cases.sort((a, b) => a.id.localeCompare(b.id));
}

function readIf(path: string): Uint8Array | null {
  return existsSync(path) ? new Uint8Array(readFileSync(path)) : null;
}

function plain(bytes: Uint8Array | null): unknown {
  return bytes === null ? undefined : toJsonValue(parseJsonBytes(bytes));
}

function run(testCase: Case): { html: string } | { error: TemplateError } {
  const optionsBytes = readIf(join(testCase.dir, 'options.json'));
  const options = (plain(optionsBytes) ?? {}) as { delimiters?: string };
  try {
    const engine = new Engine(new AstProgram({ loader: new FsLoader(testCase.dir), ...(options.delimiters ? { delimiters: options.delimiters } : {}) }));
    const assign = readIf(join(testCase.dir, 'data.json'));
    const define = plain(readIf(join(testCase.dir, 'define.json'))) as Record<string, string | { template?: string; data?: unknown; html?: string }> | undefined;
    const env = plain(readIf(join(testCase.dir, 'env.json'))) as { timezone?: string; now?: number } | undefined;
    const renderOptions: { define?: typeof define; env?: typeof env } = {};
    if (define) renderOptions.define = define;
    if (env) renderOptions.env = env;
    const html = engine.render('input.tpl', assign === null ? {} : parseJsonBytes(assign), renderOptions);
    return { html };
  } catch (error) {
    if (error instanceof TemplateError) return { error };
    if (error instanceof BindError) {
      return { error: new TemplateError({ code: error.code, template: 'input.tpl', line: 0, col: 0, offset: 0, end: 0, message: error.message }) };
    }
    throw error;
  }
}

describe('conformance cases', () => {
  for (const testCase of listCases()) {
    const expectedAst = readIf(join(testCase.dir, 'expected.ast.json'));
    const expectedHtml = readIf(join(testCase.dir, 'expected.html'));
    const expectedError = readIf(join(testCase.dir, 'expected.error.json'));
    const optionsBytes = readIf(join(testCase.dir, 'options.json'));
    const options = (plain(optionsBytes) ?? {}) as { delimiters?: string };

    if (expectedAst) {
      it(`${testCase.id} ast`, () => {
        const source = new Uint8Array(readFileSync(join(testCase.dir, 'input.tpl')));
        const ast = parse(source, 'input.tpl', options.delimiters ? { delimiters: options.delimiters } : {});
        expect(JSON.parse(JSON.stringify(ast))).toEqual(JSON.parse(new TextDecoder().decode(expectedAst)));
      });
    }
    it(`${testCase.id} render`, () => {
      const result = run(testCase);
      if (expectedError) {
        const expected = JSON.parse(new TextDecoder().decode(expectedError)) as Record<string, unknown>;
        expect('error' in result ? { code: result.error.code, template: result.error.template, line: result.error.line, col: result.error.col } : result).toEqual(expected);
      } else {
        expect('html' in result ? result.html : result.error.toObject()).toEqual(new TextDecoder().decode(expectedHtml as Uint8Array));
      }
    });
  }
});

File: packages/template-go/conformance_test.go
package template_test

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	template "github.com/polyspec/template"
	"github.com/polyspec/template/errs"
	"github.com/polyspec/template/value"
)

var casesDir = filepath.Join("..", "..", "tests", "cases")

type conformanceCase struct {
	id  string
	dir string
}

func listCases(t *testing.T) []conformanceCase {
	t.Helper()
	groups, err := os.ReadDir(casesDir)
	if err != nil {
		t.Fatal(err)
	}
	var cases []conformanceCase
	for _, group := range groups {
		if !group.IsDir() {
			continue
		}
		entries, err := os.ReadDir(filepath.Join(casesDir, group.Name()))
		if err != nil {
			t.Fatal(err)
		}
		for _, entry := range entries {
			if entry.IsDir() {
				cases = append(cases, conformanceCase{id: group.Name() + "/" + entry.Name(), dir: filepath.Join(casesDir, group.Name(), entry.Name())})
			}
		}
	}
	return cases
}

func readIf(path string) []byte {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	return data
}

func plainJSON(v value.Value) any {
	switch x := v.(type) {
	case value.SafeString:
		return x.Text
	case value.List:
		out := make([]any, len(x))
		for i, item := range x {
			out[i] = plainJSON(item)
		}
		return out
	case *value.OrderedMap:
		out := map[string]any{}
		for _, k := range x.Keys() {
			out[k] = plainJSON(x.MustGet(k))
		}
		return out
	}
	return v
}

func errorFields(err error) map[string]any {
	var te *errs.Error
	if errors.As(err, &te) {
		return map[string]any{"code": string(te.Code), "template": te.Template, "line": float64(te.Line), "col": float64(te.Col)}
	}
	return map[string]any{"unexpected": err.Error()}
}

func renderCase(c conformanceCase) (string, error) {
	options := template.Options{Loader: template.NewFSLoader(os.DirFS(c.dir))}
	if optionsBytes := readIf(filepath.Join(c.dir, "options.json")); optionsBytes != nil {
		var opts struct {
			Delimiters string `json:"delimiters"`
		}
		if err := json.Unmarshal(optionsBytes, &opts); err != nil {
			return "", err
		}
		options.Delimiters = opts.Delimiters
	}
	engine, err := template.NewEngine(options)
	if err != nil {
		return "", err
	}
	var assign value.Value = value.NewOrderedMap()
	if dataBytes := readIf(filepath.Join(c.dir, "data.json")); dataBytes != nil {
		if assign, err = value.ParseJSON(dataBytes); err != nil {
			var be *value.BindError
			if errors.As(err, &be) {
				return "", errs.WithoutPosition(be.Code, "input.tpl", be.Message)
			}
			return "", err
		}
	}
	renderOptions := template.RenderOptions{}
	if defineBytes := readIf(filepath.Join(c.dir, "define.json")); defineBytes != nil {
		var define map[string]json.RawMessage
		if err := json.Unmarshal(defineBytes, &define); err != nil {
			return "", err
		}
		renderOptions.Define = map[string]template.DefineInput{}
		for id, raw := range define {
			var path string
			if err := json.Unmarshal(raw, &path); err == nil {
				renderOptions.Define[id] = template.DefineInput{Template: path}
				continue
			}
			var entry struct {
				Template string  `json:"template"`
				Data     any     `json:"data"`
				HTML     *string `json:"html"`
			}
			if err := json.Unmarshal(raw, &entry); err != nil {
				return "", err
			}
			renderOptions.Define[id] = template.DefineInput{Template: entry.Template, Data: entry.Data, HTML: entry.HTML}
		}
	}
	if envBytes := readIf(filepath.Join(c.dir, "env.json")); envBytes != nil {
		env := &template.Env{Timezone: "Z"}
		var raw struct {
			Timezone string  `json:"timezone"`
			Now      float64 `json:"now"`
		}
		if err := json.Unmarshal(envBytes, &raw); err != nil {
			return "", err
		}
		if raw.Timezone != "" {
			env.Timezone = raw.Timezone
		}
		env.Now = raw.Now
		renderOptions.Env = env
	}
	return engine.Render("input.tpl", assign, renderOptions)
}

func TestConformanceCases(t *testing.T) {
	for _, c := range listCases(t) {
		t.Run(c.id, func(t *testing.T) {
			if expectedAST := readIf(filepath.Join(c.dir, "expected.ast.json")); expectedAST != nil {
				source, err := os.ReadFile(filepath.Join(c.dir, "input.tpl"))
				if err != nil {
					t.Fatal(err)
				}
				parseOptions := template.ParseOptions{}
				if optionsBytes := readIf(filepath.Join(c.dir, "options.json")); optionsBytes != nil {
					var opts struct {
						Delimiters string `json:"delimiters"`
					}
					_ = json.Unmarshal(optionsBytes, &opts)
					parseOptions.Delimiters = opts.Delimiters
				}
				tree, err := template.Parse(source, "input.tpl", parseOptions)
				if err != nil {
					t.Fatalf("parse: %v", err)
				}
				var got, want any
				data, _ := json.Marshal(tree)
				_ = json.Unmarshal(data, &got)
				_ = json.Unmarshal(expectedAST, &want)
				if !reflect.DeepEqual(got, want) {
					t.Errorf("ast differs\n got %s\nwant %s", data, expectedAST)
				}
			}
			html, err := renderCase(c)
			if expectedError := readIf(filepath.Join(c.dir, "expected.error.json")); expectedError != nil {
				var want map[string]any
				_ = json.Unmarshal(expectedError, &want)
				if err == nil {
					t.Fatalf("expected error %v, got output %q", want, html)
				}
				if got := errorFields(err); !reflect.DeepEqual(got, want) {
					t.Errorf("error got %v, want %v", got, want)
				}
				return
			}
			if err != nil {
				t.Fatalf("render: %v", err)
			}
			want := string(readIf(filepath.Join(c.dir, "expected.html")))
			if html != want {
				t.Errorf("html differs\n got %q\nwant %q", html, want)
			}
		})
	}
}
